### PR TITLE
qtcreator-qt6: init at 9.0.2

### LINKED ISSUE #119959: Qtcreator's clang plugins does not exist in unstable channel.
**Describe the bug**
qtcreator from unstable channel doesn't come with clang plugins even though default value of withClangPlugins is true. This problem doesn't exist in 20.09 channel.

**To Reproduce**
1. Download qtcreator from unstable channel. You can look which plugins are loaded from Help -> "About Plugins".

**Expected behavior**
Clang plugins namely ClangCodeModel, ClangFormat and ClangTools should be loaded.

**Notify maintainers**
@akaWolf

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.23-1-MANJARO, Manjaro Linux, noversion`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(ati): `"nixgl, nixpkgs-21.05pre283529.8389dcb67d9, nixpkgs-2009-20.09.3866.9a1672105db, nixpkgs-unstable-21.05pre283529.8389dcb67d9"`
 - nixpkgs: `/home/ati/.nix-defexpr/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: qtcreator
# a list of nixos modules affected by the problem
module:
```

### LINKED ISSUE #173751: qtcreator outdated
###### Checklist

- [x] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name
`nix search` name: **qtcreator**

current version: 5.0.3
desired version: 7.0.1

###### Notify maintainers

maintainers: @akaWolf 

###### Note for maintainers

Please tag this issue in your PR.

### LINKED ISSUE #197006: qtcreator is too old
###### Project name
`nix search` name:
qtcreator
current version: 5.0.3
desired version: 8.0.1

###### Notify maintainers

maintainers: @akaWolf 

###### Note for maintainers

qtcreator 8.0.1 needs qt6, but it add many features, included ui promote. So I think maybe it should be update?

### LINKED ISSUE #219735: Update request: qtcreator 5.0.3 → 9.0.2
- Package name: qtcreator
- Latest released version: 9.0.2

- Current version on the unstable channel: 5.0.3
- Current version on the stable/release channel: 5.0.3

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

@akaWolf

-----

Note for maintainers: Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by wineee**
###### Description of changes

Remove qtcreator(qt5) when  this is verified to be stable to use

Superseded：https://github.com/NixOS/nixpkgs/pull/189504
fix https://github.com/NixOS/nixpkgs/issues/173751
fix https://github.com/NixOS/nixpkgs/issues/197006
fix https://github.com/NixOS/nixpkgs/issues/219735
fix https://github.com/NixOS/nixpkgs/issues/119959

The following issues remain unresolved：

- https://github.com/NixOS/nixpkgs/issues/59580
- https://github.com/NixOS/nixpkgs/issues/59581

```
                                                                            
CMake Warning at cmake/QtCreatorDocumentation.cmake:187 (message):
  No qdoc binary found: No documentation targets were generated             
Call Stack (most recent call first):                                        
  cmake/QtCreatorDocumentation.cmake:285 (qdoc_build_qdocconf_file)         
  doc/CMakeLists.txt:38 (add_qtc_documentation)                             
                                                                            
                                                                            
-- The following packages have been found:

 * Qt6CoreTools (required version >= 6.4.3)
 * Qt6Core (required version >= 6.2.0)
 * Qt6Concurrent
 * OpenGL
 * PkgConfig
 * XKB (required version >= 0.5.0), XKB API common to servers and clients., <http://xkbcommon.org>
 * Vulkan
 * Qt6GuiTools (required version >= 6.4.3)
 * Qt6DBusTools (required version >= 6.4.3)
 * Qt6Gui
 * Qt6Network
 * Qt6WidgetsTools (required version >= 6.4.3)
 * Qt6PrintSupport
 * Qt6QmlTools (required version >= 6.4.3)
 * Qt6Qml
 * Qt6Sql
 * Qt6Xml
 * Qt6Core5Compat
 * Qt6LinguistTools
 * WrapVulkanHeaders
 * Qt6DesignerComponentsPrivate
 * Qt6Quick
 * Qt6QuickWidgets
 * Qt6Help
 * Qt6SerialPort
 * Qt6Svg
 * Qt6ToolsTools (required version >= 6.4.3)
 * Qt6Tools
 * LLVM
 * Clang
 * Qt6Test
 * Qt6SvgWidgets
 * Qt6WebEngineCoreTools (required version >= 6.4.3)
 * Qt6WebEngineWidgets
 * yaml-cpp
 * Googletest, Google Testing and Mocking Framework, <https://github.com/google/googletest>
 * Qt6Quick3DTools (required version >= 6.4.3)
 * Qt6ShaderToolsTools (required version >= 6.4.3)
 * Qt6ShaderTools (required version >= 6.4.3)
 * Qt6Quick3DRuntimeRender (required version >= 6.4.3)
 * Qt6Quick3D
 * Qt6Quick3DAssetImport
 * Qt6Quick3DParticles
 * Qt6Quick3DAssetUtils
 * Qt6QuickTimeline
 * Threads
 * Qt6
 * Qt5
 * elfutils, a collection of utilities and libraries to read, create and modify ELF binary files, <https://sourceware.org/elfutils/>
 * Zstd
 * LibRustcDemangle, Demangling for Rust symbols, written in Rust., <https://github.com/alexcrichton/rustc-demangle>
   Demangling of Rust symbols
 * Python3

-- The following packages have not been found:

 * litehtml

-- The following features have been enabled:

 * Build documentation
 * Library 3rd_cplusplus
 * Library KSyntaxHighlighting
 * Library AdvancedDockingSystem
 * Library Aggregation
 * Library ExtensionSystem
 * Library Utils
 * Library LanguageUtils
 * Library CPlusPlus
 * Library Modeling
 * Library QmlJS
 * Library QmlDebug
 * Library QmlEditorWidgets
 * Library GLSL
 * Library LanguageServerProtocol
 * Library SqliteC, with CONDITION Qt5_VERSION VERSION_GREATER_EQUAL 6.2.0
 * Library Sqlite, with CONDITION Qt5_VERSION VERSION_GREATER_EQUAL 6.2.0
 * Library Tracing, with CONDITION TARGET Qt6::ShaderTools AND TARGET Qt5::Quick
 * Library designerintegrationv2
 * Library ProParser
 * Library shared_help
 * Plugin Core
 * Plugin TextEditor
 * Plugin SerialTerminal, with CONDITION TARGET Qt5::SerialPort
 * Plugin HelloWorld
 * Plugin ImageViewer
 * Plugin Marketplace
 * Plugin UpdateInfo
 * Plugin Welcome
 * Plugin BinEditor
 * Plugin CodePaster
 * Plugin DiffEditor
 * Plugin EmacsKeys
 * Plugin Macros
 * Plugin ProjectExplorer
 * Plugin SilverSearcher
 * Plugin Bookmarks
 * Plugin CppEditor
 * Plugin Help, with CONDITION TARGET Qt5::Help
 * QtWebEngine help viewer, with CONDITION BUILD_HELPVIEWERBACKEND_QTWEBENGINE AND TARGET Qt5::WebEngineWidgets
 * litehtml help viewer, with CONDITION TARGET qlitehtml
 * Plugin ResourceEditor
 * Plugin Nim
 * Plugin Conan
 * Plugin ClassView
 * Plugin GLSLEditor
 * Plugin ModelEditor
 * Plugin QtSupport
 * Plugin Todo
 * Plugin VcsBase
 * Plugin Bazaar
 * Plugin Beautifier
 * Plugin ClangFormat, with CONDITION TARGET clang-cpp AND LLVM_PACKAGE_VERSION VERSION_GREATER_EQUAL 10.0.0 AND ( QTC_CLANG_BUILDMODE_MATCH OR CLANGTOOLING_LINK_CLANG_DYLIB )
 * Plugin ClearCase
 * Plugin CVS
 * Plugin Designer, with CONDITION TARGET Qt5::DesignerComponents AND TARGET Qt5::Designer
 * Plugin Docker
 * Plugin FakeVim
 * Plugin GenericProjectManager
 * Plugin Git
 * Plugin Mercurial
 * Plugin MesonProjectManager
 * Plugin Perforce
 * Plugin QmakeProjectManager
 * Plugin QmlJSTools
 * Plugin QmlProjectManager, with CONDITION TARGET Qt5::QuickWidgets
 * Plugin ScxmlEditor
 * Plugin Subversion
 * Plugin CompilationDatabaseProjectManager
 * Plugin LanguageClient
 * Plugin CMakeProjectManager
 * Plugin Debugger
 * Plugin Coco
 * Plugin GitLab
 * Plugin Android
 * Plugin AutoTest
 * Plugin AutotoolsProjectManager
 * Plugin BareMetal
 * Plugin ClangCodeModel
 * Plugin ClangTools, with CONDITION TARGET yaml-cpp
 * Plugin Cppcheck
 * Plugin IncrediBuild
 * Plugin Ios
 * Plugin Python
 * Plugin QmlJSEditor
 * Plugin QmlPreview, with CONDITION TARGET QmlProjectManager
 * Plugin QmlProfiler, with CONDITION TARGET Tracing AND TARGET Qt6::ShaderTools
 * Plugin RemoteLinux
 * Plugin Valgrind
 * Plugin PerfProfiler, with CONDITION TARGET Tracing
 * Plugin QbsProjectManager
 * Plugin CtfVisualizer, with CONDITION TARGET Tracing
 * Plugin Squish
 * Plugin Boot2Qt
 * Plugin QmlDesigner, with CONDITION Qt5_VERSION VERSION_GREATER_EQUAL 6.2.0 AND TARGET Qt5::QuickWidgets AND TARGET Qt5::Svg
 * Plugin assetexporterplugin, with CONDITION TARGET QmlDesigner
 * Plugin componentsplugin, with CONDITION TARGET QmlDesigner
 * Plugin qmlpreviewplugin, with CONDITION TARGET QmlDesigner
 * Plugin qtquickplugin, with CONDITION TARGET QmlDesigner
 * Plugin StudioPlugin, with CONDITION TARGET QmlDesigner
 * Plugin StudioWelcome, with CONDITION TARGET Qt5::QuickWidgets AND TARGET QmlDesigner
 * Plugin Qnx
 * Plugin WebAssembly
 * Plugin McuSupport
 * Library qml2puppet_static
 * Qt Quick 3D support, with CONDITION TARGET Qt5::Quick3D
 * Qt Quick 3D asset import, with CONDITION TARGET Qt5::Quick3DAssetImport
 * Qt Quick 3D particles, with CONDITION TARGET Qt5::Quick3DParticles
 * Qt Quick 3D asset utils, with CONDITION TARGET Qt5::Quick3DAssetUtils
 * Library sdktoolLib
 * Library perfparser_lib
 * Include developer documentation

-- The following features have been disabled:

 * Build online documentation
 * Build tests
 * Build with sanitize, SANITIZE_FLAGS=''
 * Build with Crashpad
 * Library Nanotrace
 * Build Qbs
 * Native WebKit help viewer, with CONDITION FWWebKit AND FWAppKit
 * ProjectStorage
 * Build with QmlDom
 * multilanguage-support in qml2puppet, with CONDITION TARGET QtCreator::multilanguage-support

```

1. litehtml use `find_dependency(gumbo)` but gumbo only provide pkg-config file.
https://github.com/archlinux/svntogit-packages/blob/1cc60cd7bed3034608eb137ff82761e7ec88acb5/trunk/PKGBUILD#LL29C48-L29C56
https://github.com/NixOS/nixpkgs/pull/223361
cc @fgaz 

2. qdoc （Should be provided by qttools）not found，this also caused qtdoc to not be packaged correctly
https://github.com/NixOS/nixpkgs/pull/223271

3. optional runtime dependencies：
https://github.com/NixOS/nixpkgs/pull/223254

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/litehtml/default.nix [hunk 1]
```diff
@@ -16,6 +16,12 @@ stdenv.mkDerivation rec {
     hash = "sha256-9571d3k8RkzEpMWPuIejZ7njLmYstSwFUaSqT3sk6uQ=";
   };
 
+  # Don't search for non-existant gumbo cmake config
+  postPatch = ''
+    substituteInPlace cmake/litehtmlConfig.cmake \
+      --replace "find_dependency(gumbo)" ""
+  '';
+
```
Comment (fgaz) [inline pkgs/development/libraries/litehtml/default.nix:24]: Why is this here?

Comment (wineee) [reply pkgs/development/libraries/litehtml/default.nix:24]: Gumbo don't provide cmake config, `find_dependency(gumbo)` will cause failure

Archlinux does something similar： https://github.com/archlinux/svntogit-packages/blob/1cc60cd7bed3034608eb137ff82761e7ec88acb5/trunk/PKGBUILD#LL29C48-L29C56

Comment (fgaz) [reply pkgs/development/libraries/litehtml/default.nix:24]: I mean why is this in a pr that doesn't touch anything in the litehtml closure

Also it isn't failing: https://hydra.nixos.org/build/213914327

Comment (wineee) [reply pkgs/development/libraries/litehtml/default.nix:24]: qtcreator use cmake to search litehtml, need this to ensure litehtml can be found. 
```
-- The following packages have not been found:

 * litehtml
```
Here should be a separate pr： https://github.com/NixOS/nixpkgs/pull/223361

Comment (wineee) [reply pkgs/development/libraries/litehtml/default.nix:24]: > Also it isn't failing: [hydra.nixos.org/build/213914327](https://hydra.nixos.org/build/213914327)

This does not affect the building of litehtml itself, but the software that depends on litehtml, we can also substitute in `postInstall`

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -18656,6 +18656,12 @@ with pkgs;
     inherit (linuxPackages) perf;
   };
 
+  qtcreator_qt6 = qt6Packages.callPackage ../development/tools/qtcreator/qt6.nix {
```
Comment (wegank) [inline pkgs/top-level/all-packages.nix:18659]:
```suggestion
  qtcreator-qt6 = qt6Packages.callPackage ../development/tools/qtcreator/qt6.nix {
```

File: pkgs/development/tools/qtcreator/qt6.nix [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+, perf
+}:
+
+stdenv.mkDerivation rec {
+  pname = "qtcreator";
+  version = "9.0.2";
+  baseVersion = builtins.concatStringsSep "." (lib.take 2 (builtins.splitVersion version));
+
+  src = fetchurl {
+    url = "https://download.qt.io/official_releases/${pname}/${baseVersion}/${version}/qt-creator-opensource-src-${version}.tar.xz";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/qtcreator/qt6.nix:35]:
```suggestion
    url = "https://download.qt.io/official_releases/qtcreator/${lib.versions.majorMinor version}/${version}/qt-creator-opensource-src-${version}.tar.xz";
```

File: pkgs/development/tools/qtcreator/qt6.nix [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+, llvmPackages
+, rustc-demangle
+, elfutils
+, perf
+}:
+
+stdenv.mkDerivation rec {
+  pname = "qtcreator";
+  version = "9.0.2";
+  baseVersion = builtins.concatStringsSep "." (lib.take 2 (builtins.splitVersion version));
```
Comment (SuperSandro2000) [inline pkgs/development/tools/qtcreator/qt6.nix:32]:
```suggestion
```

**Review by SuperSandro2000 — COMMENTED** [review]
otherwise LGTM

**Comment by wineee** [comment]
@ofborg build qtcreator

**Review by wegank — APPROVED** [review]
